File: src/app/android/app/src/main/java/com/cyberacy/app/models/entities/Meeting.kt
package com.cyberacy.app.models.entities

import android.util.Log
import com.cyberacy.app.models.services.ApiConnection
import com.google.gson.annotations.SerializedName
import retrofit2.HttpException
import retrofit2.await
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter

class Meeting(
    @SerializedName("id")
    val id: Int,

    @SerializedName("name")
    val name: String,

    @SerializedName("object")
    val meetingObject: String,

    @SerializedName("description")
    val description: String?,

    @SerializedName("date_start")
    val dateStart: LocalDateTime,

    @SerializedName("nb_time")
    val nbTime: Double,

    @SerializedName("price_excl")
    val priceExcl: Double,

    @SerializedName("vta_rate")
    val rateVTA: Double,

    @SerializedName("is_aborted")
    val isAborted: Boolean,

    @SerializedName("reason_aborted")
    val reasonAborted: Boolean,

    @SerializedName("nb_place")
    val nbPlace: Int,

    @SerializedName("nb_place_vacant")
    val nbPlaceVacant: Int,

    @SerializedName("address_street")
    val addressStreet: String?,

    @SerializedName("link_twitch")
    val linkTwitch: String?,

    @SerializedName("link_youtube")
    val linkYoutube: String?,

    @SerializedName("town_code_insee")
    val townCodeInsee: String?,

    @SerializedName("latitude")
    val latitude: Double?,

    @SerializedName("longitude")
    val longitude: Double?,

    @SerializedName("is_participate")
    val isParticipated: Boolean = false,

    @SerializedName("town_name")
    val townName: String?,

    @SerializedName("town")
    val town: Town?,
) {

    fun getMonthPrefix(): String {
        val listMonthPrefix = listOf(
            "JANV",
            "FEV",
            "MARS",
            "AVR",
            "MAI",
            "JUIN",
            "JUIL",
            "AOÛT",
            "SEPT",
            "OCT",
            "NOV",
            "DEC"
        )
        return listMonthPrefix[dateStart.monthValue - 1]
    }

    fun getPriceStr(): String {
        return if (priceExcl <= 0.00) {
            "gratuit"
        } else {
            val price = priceExcl + (priceExcl * (rateVTA / 100))
            "$price €"
        }
    }

    fun getPosition(): String {
        return if (addressStreet == null || townCodeInsee == null || addressStreet.isEmpty() || townCodeInsee.isEmpty()) {
            if (linkYoutube != null) {
                "Youtube"
            } else if (linkTwitch != null) {
                "Twitch"
            } else {
                "Inconnu"
            }
        } else {
            addressStreet
        }
    }

    fun getNbPlaceStr(): String {
        return if(nbPlaceVacant < 0) {
            "Illimité"
        } else if(nbPlaceVacant == 0) {
            "Complet"
        } else if(nbPlaceVacant in 1..10) {
            "Plus que $nbPlaceVacant places"
        } else {
            "$nbPlaceVacant places restantes"
        }
    }

    fun getAddressFull(): String {
        var addressFull = addressStreet ?: ""
        if(town != null) {
            addressFull += ", ${town.zip_code} ${town.name}"
        }
        return addressFull
    }

    fun getDateMeeting(): String{
        val formatterDate: DateTimeFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm")
        val formatterTime: DateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm")
        val today: LocalDate = LocalDateTime.now().atZone(ZoneId.systemDefault()).toLocalDate()
        val dateMeeting: LocalDate = dateStart.atZone(ZoneId.systemDefault()).toLocalDate()
        return if(today.isEqual(dateMeeting)) {
            "Aujourd'hui à ${formatterTime.format(dateStart)}"
        } else {
            formatterDate.format(dateStart)
        }
    }

    fun getTimeStr(): String{
        if(nbTime <= 0) {
            return "Indéterminée"
        }
        val nbTotalMinutes = nbTime * 60
        val nbHour = nbTime.toInt()
        val nbMinutes = (nbTotalMinutes - (nbHour * 60)).toInt()
        return "${nbHour}h$nbMinutes"
    }

    fun getTimeEnd(): LocalDateTime{
        return dateStart.plusMinutes((nbTime * 60).toLong())
    }

    fun getDescriptionStr(): String{
        return description ?: "Aucune description"
    }

    fun reservationIsAvailable(): Boolean {
        return (dateStart > LocalDateTime.now() && nbPlaceVacant != 0)
    }

    companion object Service {

        suspend fun getMeetings(partyId: Int): List<Meeting> {
            try {
                val response: ResponseAPI<List<Meeting>?> = ApiConnection.connection().getMeeting(
                    idPoliticalParty = partyId,
                    mine = false,
                    includeCompleted = true,
                    includeFinished = true
                ).await()
                return response.data ?: emptyList()
            } catch (e: HttpException) {
                throw e
            }
        }

        suspend fun getMeetingById(id: Int): Meeting? {
            try {
                val response: ResponseAPI<Meeting?> = ApiConnection.connection().getMeetingById(id).await()
                return response.data
            } catch (e: HttpException) {
                throw e
            }
        }

        suspend fun participatedToMeeting(id: Int): Unit {
            try {
                val response: ResponseAPI<Unit> = ApiConnection.connection().participateToMeeting(id).await()
                return
            } catch (e: HttpException) {
                throw e
            }
        }

    }

}

File: src/client/flutter/lib/models/entities/meeting.dart
import 'package:bo_cyberacy/models/entities/town.dart';
import 'package:bo_cyberacy/models/enums/state_progress.dart';

class Meeting {

  int id = -1;
  String name;
  DateTime dateStart;
  String? description;
  String? object;
  double nbTime;
  bool isAborted;
  int nbPlaceVacant;
  int nbPlace;
  String? addressStreet;
  int idPoliticalParty;
  String townCodeInsee;
  double rateVAT;
  double priceExcl;
  String? townName;
  bool isParticipated;
  String? linkTwitch;
  String? linkYoutube;
  double? latitude;
  double? longitude;
  Town? town;

  Meeting({
    this.id = -1,
    required this.name,
    required this.dateStart,
    required this.nbTime,
    this.isAborted = false,
    this.nbPlaceVacant = 0,
    this.nbPlace = -1,
    this.addressStreet,
    required this.idPoliticalParty,
    required this.townCodeInsee,
    required this.rateVAT,
    required this.priceExcl,
    this.townName,
    this.isParticipated = false,
    this.linkTwitch,
    this.linkYoutube,
    this.latitude,
    this.longitude,
    this.description,
    this.object,
    this.town
  });

  Meeting.fromJson(Map<String, dynamic> json)
      : id = json["id"],
        name = json["name"],
        dateStart = DateTime.parse(json["date_start"] ?? "2022-05-05"),
        nbTime = double.parse(json["nb_time"] ?? "0.0"),
        isAborted = json["is_aborted"],
        nbPlaceVacant = json["nb_place_vacant"] ?? -1,
        nbPlace = json["nb_place"] ?? -1,
        addressStreet = json["address_street"],
        idPoliticalParty = json["id_political_party"],
        townCodeInsee = json["town_code_insee"],
        rateVAT = double.parse(json["vta_rate"] ?? "0.0"),
        priceExcl = double.parse(json["price_excl"] ?? "0.0"),
        townName = json["town_name"],
        isParticipated = json["is_participate"],
        linkTwitch = json["link_twitch"],
        linkYoutube = json["link_youtube"],
        latitude = double.tryParse(json["latitude"] ?? ""),
        longitude = double.tryParse(json["longitude"] ?? ""),
        description = json["description"],
        object = json["object"] {
    if(json["town"] != null) {
      Map<String, dynamic> townJson = json["town"];
      town = Town.fromJson(townJson);
    }
  }

  Object toJson() {
    return {
      "name": name,
      "date_start": dateStart.toString(),
      "nb_time": nbTime.toString(),
      "nb_place": nbPlace.toString(),
      "street_address": addressStreet,
      "id_political_party": idPoliticalParty.toString(),
      "town_code_insee": townCodeInsee,
      "vta_rate": rateVAT.toString(),
      "price_excl": priceExcl.toString(),
      "link_twitch": linkTwitch,
      "link_youtube": linkYoutube,
      "description": description,
      "object": object
    };
  }

  String getFullAddress() {
    String fullAddress = addressStreet ?? "-- aucune adresse --";
    if(townName != null) {
      fullAddress += ", $townName";
    }
    return fullAddress;
  }

  StateProgress getStateProgress() {
    DateTime today = DateTime.now();
    int totalMinutes = (nbTime * 60).toInt();
    DateTime dateEnd = dateStart.add(Duration(minutes: totalMinutes));
    if(today.isAfter(dateEnd)) {
      return StateProgress.passed;
    } else if (today.isBefore(dateStart)) {
      return StateProgress.coming;
    } else {
      return StateProgress.inProgress;
    }
  }

}